win32ss/user/user32/windows/cursoricon.c

```diff
@@ -7,11 +7,376 @@
+    {
+        ERR("Not Enough working room\n");
+        goto exit_error;
+    }
+    else
+    {
+        wcscat(lpTempPathBuffer, L"bmpout.ico1");
+    }
+
+    fp = _wfopen(lpTempPathBuffer, L"wb");
+    if (!fp)
+    {
+        ERR("Temp Directory Path File Open Failed\n");
+        goto exit_error;
+    }
+    else
+    {
+        fclose(fp);
+    }
+
+    _wremove(lpTempPathBuffer);
+    lpTempPathBuffer[wcslen(lpTempPathBuffer) - 1] = 0;
+
+    return out;
+exit_error:
+    return 0;
+}
+
+/* libpng defines */
+#define png_infopp_NULL (png_infopp)NULL
+#define PNG_BYTES_TO_CHECK 4
+
+/* libpng helpers */
+typedef struct {
+    png_bytep buffer;
+    png_uint_32 bufsize;
+    png_uint_32 current_pos;
+} MEMORY_READER_STATE;
+
+struct png_wrapper
+{
+    const char *buffer;
+    size_t size, pos;
+};
+
+/* This function will be used for reading png data from array */
+static void read_data_memory(png_structp png_ptr, png_bytep data, png_uint_32 length) 
+{
+    MEMORY_READER_STATE *f = png_get_io_ptr(png_ptr);
+    if (length > (f->bufsize - f->current_pos))
+        png_error(png_ptr, "read error in read_data_memory (loadpng)");
+    memcpy(data, f->buffer + f->current_pos, length);
+    f->current_pos += length;
+}
+
+void PNGtoBMP(_In_ LPBYTE pngbits, _In_ DWORD filesize, _Out_ LPBYTE outbits)
+{
+    int is_png;
+    BITMAPINFOHEADER info = { 0 };
+    CURSORICONFILEDIR cifd = { 0 };
+    int bpp = 0;
+    int image_size = 0;
+    FILE * fp;
+    CHAR lpTempPathBuffer[MAX_PATH + 1];
+    MEMORY_READER_STATE memory_reader_state;
+    png_bytep mem_read_ptr = (png_bytep)&memory_reader_state;
+    png_uint_32 width, height, channels;
+    int bit_depth, color_type, interlace_type;
+    png_bytep* row_pointers;
+    int width1, height1, size, rowbytes, pos, stride;
+    LPBYTE data;
+
+    if (!pngbits)
+        return;
+
+    is_png = !png_sig_cmp((png_const_bytep)pngbits, 0, 8);
+
+    TRACE("is_png %d and filesize %d\n", is_png, filesize);
+
+    if (!is_png)
+        return;
+
+    png_structp png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+    if (!png_ptr)
+    {
+        ERR("png_create_read_struct error\n");
+        return;
+    }
+
+    png_infop info_ptr = png_create_info_struct(png_ptr);
+    if (!info_ptr)
+    {
+        ERR("png_create_info_struct error\n");
+        png_destroy_read_struct(&png_ptr, (png_infopp)NULL, (png_infopp)NULL);
+        return;
+    }
+
+    png_infop end_info = png_create_info_struct(png_ptr);
+    if (!end_info)
+    {
+        ERR("png end_info error\n");
+        png_destroy_read_struct(&png_ptr, &info_ptr, (png_infopp)NULL);
+        return;
+    }
+
+    // png_source is array which has png data
+    memory_reader_state.buffer = (png_bytep)pngbits;
+    memory_reader_state.bufsize = filesize;
+    memory_reader_state.current_pos = PNG_BYTES_TO_CHECK;
+
+    // set our own read_function
+    png_set_read_fn(png_ptr, mem_read_ptr, read_data_memory);
```

Comment:
Here, Line 226.

The prototype of `png_set_read_fn` is:
```c
PNG_EXPORT(78, void, png_set_read_fn, (png_structrp png_ptr, png_voidp io_ptr,
    png_rw_ptr read_data_fn));
```
The 1st parameter `png_ptr` is correct.
The 2st parameter `mem_read_ptr` should be compatible to `png_voidp`.
The 3rd parameter `read_data_memory` should be compatible to `png_rw_ptr`.

`png_voidp` is same as `void *`.
`png_rw_ptr` is defined as:

```c
typedef PNG_CALLBACK(void, *png_rw_ptr, (png_structp, png_bytep, size_t));
```

However your prototype is:

```c
static void read_data_memory(png_structp png_ptr, png_bytep data, png_uint_32 length)
```

Type `size_t` and `png_uint_32` are mismatching on x64.